File: devardalamarraetal2025_iconicity/generate_prompts.R
### convert psycholinguistic iconicity rating data into a zipped JSONL file suitable for fine-tuning and evaluating large language models ###

# input:   exp1.csv         
# output:  prompts.jsonl.zip 

# each JSON object contains:
#   - "text":           Full prompt (instructions + trial-by-trial stimuli with human ratings marked as <<rating>>)
#   - "experiment":     Experiment identifier ("exp1")
#   - "participant_id": Participant ID
#   - "group":          Participant's group (1=L1 Ita, 2=L2 Ita)
#   - first_language:   Participant's first language
#   - second_language:  Participant's second language
#   - lexita_self_cefr: Participant's score in LexIta according to CEFR levels
#   - lexita_score:     Participant's score in LexIta


library(dplyr)
library(jsonlite)

# Resolve paths from this script's location so it runs from any working
# directory and always writes inside its own study folder. Previously the
# script used bare relative paths and only worked with the study folder as CWD.
.args <- commandArgs(trailingOnly = FALSE)
SCRIPT_DIR <- dirname(sub("^--file=", "", .args[grep("^--file=", .args)]))
if (length(SCRIPT_DIR) == 0 || !nzchar(SCRIPT_DIR)) SCRIPT_DIR <- getwd()
SCRIPT_DIR <- normalizePath(SCRIPT_DIR)

## import data
data <- read.csv(file.path(SCRIPT_DIR, "processed_data", "exp1.csv"), stringsAsFactors = FALSE)
data$response <- round(data$response) #to shorten answers with decimals

# quick check: print column names and number of rows to verify correct loading
cat("Columns:", paste(colnames(data), collapse = ", "), "\n")
cat("Rows:", nrow(data), "\n\n")


## define the instructions text: the fixed preamble shown at the start of every participant's prompt
# as in the original study, instructions are given in Italian and then English

instructions <- "Nel seguente questionario ti verr\u00e0 chiesto di valutare circa 250 parole della lingua italiana in base alla loro ICONICIT\u00c0. Prima di iniziare, leggi attentamente le seguenti istruzioni. Alla pagina successiva troverai le stesse istruzioni in lingua inglese, nel caso l'italiano non sia la tua prima lingua. Alcune parole della lingua italiana ricordano il loro significato (ad esempio, per come 'suonano'). Queste parole sono definite 'iconiche'. Una persona potrebbe essere in grado di indovinare il significato delle parole iconiche anche se non conosce l'italiano. In questo studio, ti chiediamo di valutare quanto sono iconiche alcune parole su una scala da 1 a 7. Seleziona il punteggio 1 se la parola non \u00e8 affatto iconica, ovvero se non percepisci un legame tra il suono della parola e il suo significato. Seleziona il punteggio 7 se la parola \u00e8 molto iconica, ovvero se percepisci un legame molto forte tra il suono della parola e il suo significato. Tra le parole ad iconicit\u00e0 molto alta, che potrebbero ricevere il punteggio 7, troviamo 'boom', 'crash', 'bang' eccetera. Tra le parole a bassa iconicit\u00e0, che potrebbero ricevere un punteggio molto basso, troviamo 'ragione', 'manico', 'pensione'. Per fare un altro esempio, una parola come 'piccino' dovrebbe avere un valore di iconicit\u00e0 pi\u00f9 alto di 'azzurro'. Prima di valutare la parola, \u00e8 importante che la parola venga pronunciata ad alta voce, anche pi\u00f9 volte, e che ci si concentri sul suo significato. Cerca di concentrarti sul significato dell'intera parola, piuttosto che scomporla in pi\u00f9 parti. Per esempio, per valutare una parola come 'FONDAMENTA' pensa alle strutture interrate di sostegno di un edificio, piuttosto che alle parole 'FONDA' e 'MENTA', e concentrati su quanto il significato dell'intera parola sia simile alla pronuncia della parola 'FONDAMENTA'. Se non conosci il significato o la pronuncia di una parola, puoi selezionare la casella apposita. Ricorda di pronunciare la parola e di pensare attentamente al significato di ogni parola che valuterai. Nell'esprimere il tuo giudizio ricordati di utilizzare tutti i valori della scala e di non preoccuparti quanto spesso usi un certo valore. Per ogni parola dovrai scegliere il numero che meglio indica il tuo giudizio. Incominciamo? In this task you will be rating about 250 Italian words on their ICONICITY. Please read the following instructions very carefully as they are important for doing this task. Some Italian words sound like what they mean. These words are iconic. You might be able to guess the meaning of such a word even if you did not know Italian. Some words that might be rated high in iconicity are 'boom', 'crash', 'bang' because they sound very much like what they mean. Some words that might be rated low in iconicity are 'pensione', 'manico', 'ragione'. As another example, a word like 'piccino' should receive a higher iconicity value than 'azzurro'. In this task, you are going to rate words for how iconic they are. You will rate each word on a scale from 1 to 7. A rating of 1 indicates that the word is not at all iconic and does not at all sound like what it means. 7 indicates that the word is high in iconicity and sounds very much like what it means. It is important that you say the word out loud to yourself, and that you think about its meaning. If you are unsure of the meaning or the pronunciation of a word, you have the option of skipping it. Try to focus on the word meaning of the whole word, rather than decomposing it into parts. For example, when rating 'FONDAMENTA' think of the foundation of the house rather than 'FONDA' and 'MENTA' and rate how well the whole meaning relates to the sound of the whole word 'FONDAMENTA'. Please remember to say the word to yourself and to think about the meaning of each word. Ready to start?"


## approximate token count (to stay within the 32K token limit per participant)
# rough estimate: 1 token is about 4 characters (standard heuristic for European languages)

count_tokens_approx <- function(text) {
  nchar(text) / 4
}

TOKEN_LIMIT <- 32000  # maximum tokens allowed per participant prompt


## create sorted list of unique participant IDs

participant_ids <- sort(unique(data$participant_id))
cat("Number of participants:", length(participant_ids), "\n\n")


## define output paths

output_file <- file.path(SCRIPT_DIR, "prompts.jsonl")      # temporary unzipped file
zip_file    <- file.path(SCRIPT_DIR, "prompts.jsonl.zip")  # final output


## format trial  

format_trial <- function(trial_idx, stimulus, response) {
  paste0(
    "Trial ", trial_idx, ". ",
    "La parola italiana è: '", stimulus, "'. ",
    "Quanto è iconica questa parola? ",
    "1 (Per niente iconica) 2 3 4 5 6 7 (Moltissimo iconica). ",
    "Hai valutato: <<", response, ">>"
  )
}

# function to print examples 

print_example_prompts <- function(df, n_participants = 2, n_trials = 5) {
  cat("\n", strrep("=", 70), "\n", sep = "")
  cat("EXAMPLE PROMPTS\n")
  cat(strrep("=", 70), "\n", sep = "")
  
  # select the first n_participants 
  sample_ids <- sort(unique(df$participant_id))[
    seq_len(min(n_participants, length(unique(df$participant_id))))
  ]
  
  # print participant header
  for (pid in sample_ids) {
    cat("\n", strrep("-", 70), "\n", sep = "")
    cat("PARTICIPANT:", pid, "\n")
    cat(strrep("-", 70), "\n\n")
    
    # print the prompt
    cat(instructions, "\n\n")

    # subset this participant's trials applying the same randomization used for prompt generation
    sub_df <- df %>% filter(participant_id == pid)
    set.seed(as.integer(sub("rater_", "", pid)))
    sub_df <- sub_df[sample(nrow(sub_df)), ]

    # print the fist _trials
    for (i in seq_len(min(n_trials, nrow(sub_df)))) {
      row <- sub_df[i, ]
      cat(format_trial(i, row$stimulus, row$response), "\n")
    }
    
    # total n of completed trials per participant
    cat("... (", nrow(sub_df), "trials total)\n")
  }
  cat("\n")
}


## build and write one JSON object per participant

con <- file(output_file, open = "w", encoding = "UTF-8")
n_truncated <- 0

for (pid in participant_ids) {

  # subset and randomize trial order 
  participant_data <- data %>% filter(participant_id == pid)

  seed_value <- as.integer(sub("rater_", "", pid))
  set.seed(seed_value)

  participant_data <- participant_data[sample(nrow(participant_data)), ]

  # build trial's rows
  trial_lines <- mapply(
    format_trial,
    trial_idx = seq_len(nrow(participant_data)),
    stimulus  = participant_data$stimulus,
    response  = participant_data$response,
    SIMPLIFY  = TRUE
  )

  trials_text <- paste(trial_lines, collapse = "\n")

  # assemble full prompt
  full_text <- paste0(instructions, "\n\n---\n\n", trials_text)

  # check token limit; if reached, trials are dropped from the end until the prompt fits
  if (count_tokens_approx(full_text) > TOKEN_LIMIT) {

    n_truncated <- n_truncated + 1
    cat("WARNING: Participant", pid, "exceeds 32K token limit. Truncating trials.\n")

    instructions_tokens <- count_tokens_approx(instructions) + 10  
    remaining_budget    <- TOKEN_LIMIT - instructions_tokens

    kept_trials <- c()
    for (trial in trial_lines) {
      trial_tokens <- count_tokens_approx(trial) + 1  # +1 for newline
      if (remaining_budget - trial_tokens < 0) break # stop if next trial won't fit
      kept_trials      <- c(kept_trials, trial)
      remaining_budget <- remaining_budget - trial_tokens
    }
    
    # rebuild the prompt with only the trials that fit
    trials_text <- paste(kept_trials, collapse = "\n")
    full_text   <- paste0(instructions, "\n\n---\n\n", trials_text)
  }
  
  # additional limit: truncate at 100,000 characters if needed (GitHub file line limit)
  if (nchar(full_text) > 100000) {
    
    cat("WARNING: Participant", pid, "exceeds 100,000 character limit. Truncating.\n")
    
    # Rebuild trial by trial, stopping before the character limit
    char_budget <- 100000 - nchar(instructions) - nchar("\n\n---\n\n") - 10  
    
    kept_trials <- c()
    for (trial in trial_lines) {
      trial_chars <- nchar(trial) + 1  # +1 for newline
      if (char_budget - trial_chars < 0) break
      kept_trials  <- c(kept_trials, trial)
      char_budget  <- char_budget - trial_chars
    }
    
    trials_text <- paste(kept_trials, collapse = "\n")
    full_text   <- paste0(instructions, "\n\n---\n\n", trials_text)
  }

  # convert the prompt to a JSON object with three fields: text, experiment, participant_id.
  json_obj <- toJSON(
    list(
      text             = full_text,
      experiment       = "exp1",
      participant_id   = as.character(pid),
      group            = as.integer(participant_data$group[1]),
      first_language   = as.character(participant_data$first_language[1]),
      second_language  = as.character(participant_data$second_language[1]),
      lexita_self_cefr = as.character(participant_data$lexita_self_cefr[1]),
      lexita_score     = as.numeric(participant_data$lexita_score[1])
    ),
    auto_unbox   = TRUE,
    ensure_ascii = FALSE
  )

  writeLines(json_obj, con = con)
}

close(con)


## zip the JSONL file

zip::zip(zip_file, files = basename(output_file), root = SCRIPT_DIR)
file.remove(output_file)
cat("Created zip archive:", zip_file, "\n")

## summary report

cat("\n=== Done ===\n")
cat("Output:                      ", zip_file, "\n")
cat("Participants written:         ", length(participant_ids), "\n")
cat("Participants truncated for token limit:       ", n_truncated, "\n")

## approximate token count for the first participant as a sanity check
sample_pid  <- participant_ids[1]
sample_data <- data %>% filter(participant_id == sample_pid)
approx_tok  <- count_tokens_approx(instructions) +
  sum(nchar(sample_data$stimulus)) / 4 * 2
cat("Approx. tokens (sample participant):", round(approx_tok), "\n")


## print example prompts for visual verification
print_example_prompts(data, n_participants = 2, n_trials = 5)

File: devardaetal2024_rating/generate_prompts.py
"""
Generate LLM prompts for de Varda et al. (2024) Cloze Probability Rating Task

This script reads the processed rating data and generates a JSONL file
with one line per participant, following the PsychLing-101 format.

Participants completed a rating task where they saw sentence fragments and a 
target word, and rated (1-5) how much they expected to see that word following
the sentence fragment.
"""

import json
import zipfile
from pathlib import Path
import pandas as pd
import numpy as np

# --- Configuration ---
SCRIPT_DIR = Path(__file__).parent
PROCESSED_DATA_DIR = SCRIPT_DIR / "processed_data"
INPATH = PROCESSED_DATA_DIR / "exp1.csv"
OUTPATH = SCRIPT_DIR / "prompts.jsonl"
ZIP_OUTPATH = SCRIPT_DIR / "prompts.jsonl.zip"

# --- Instructions (based on original experiment) ---
INSTRUCTION_TEXT = """In this task, you will read sentence fragments paired with a target word and rate how much you would expect to see that word following the sentence fragment.

For each sentence fragment and target word pair, rate on a scale from 1 to 5 how much you would expect the presented word to follow the sentence fragment. Note: we are not asking you to evaluate how plausible or sensible that word is, but rather how much you would expect to find it while reading the preceding sentence context.

Scale:
1 - Not at all expected
5 - Very much expected

"""


def format_trial_description(trial_idx: int, stimulus: str, target_word: str, rating: int) -> str:
    """
    Format a single trial into a descriptive string.
    
    Args:
        trial_idx: Trial number for display (1-indexed)
        stimulus: The sentence fragment shown to participant
        target_word: The word participant rated
        rating: The rating given (1-5)
        
    Returns:
        Formatted trial description string
    """
    rating_str = str(int(rating)) if pd.notna(rating) and not np.isnan(rating) else "NA"
    return f"Trial {trial_idx}. The sentence is: '{stimulus}'. How much would you expect to read the word '{target_word}' as the next word of this sentence fragment? 1 (Not at all) 2 3 4 5 (Very much). You rate: <<{rating_str}>>"


def print_example_prompts(df: pd.DataFrame, n_participants: int = 2, n_trials: int = 5):
    """Print nicely formatted example prompts for verification."""
    print("\n" + "=" * 70)
    print("EXAMPLE PROMPTS")
    print("=" * 70)
    
    participants = sorted(df["participant_id"].unique())[:n_participants]
    
    for pid in participants:
        sub_df = df[df["participant_id"] == pid].sort_values("trial_order").head(n_trials)
        
        print(f"\n{'─' * 70}")
        print(f"PARTICIPANT: {pid}")
        print(f"{'─' * 70}")
        print(INSTRUCTION_TEXT)
        
        for _, row in sub_df.iterrows():
            desc = format_trial_description(
                trial_idx=int(row["trial_order"]) + 1,
                stimulus=str(row["stimulus"]),
                target_word=str(row["target_word"]) if pd.notna(row["target_word"]) else "",
                rating=row["response"]
            )
            print(desc)
        
        print(f"... ({len(df[df['participant_id'] == pid])} trials total)")
        print()


def main():
    """Main prompt generation function."""
    print(f"Loading data from: {INPATH}")
    
    if not INPATH.exists():
        print(f"Error: Input file not found at {INPATH}")
        print("Please run preprocess_data.py first.")
        exit(1)
    
    df = pd.read_csv(INPATH, low_memory=False)
    
    # Get unique participants
    participants = sorted(df["participant_id"].unique())
    print(f"Processing data for {len(participants)} unique participants...")
    
    records_written = 0
    
    with OUTPATH.open("w", encoding="utf8") as fo:
        for pid in participants:
            # Get trials for this participant (sorted by presentation order)
            sub_df = df[df["participant_id"] == pid].sort_values("trial_order")
            
            # Build trial descriptions (using trial_order for correct presentation sequence)
            trials = []
            for _, row in sub_df.iterrows():
                desc = format_trial_description(
                    trial_idx=int(row["trial_order"]) + 1,  # 1-indexed for display
                    stimulus=str(row["stimulus"]),
                    target_word=str(row["target_word"]) if pd.notna(row["target_word"]) else "",
                    rating=row["response"]
                )
                trials.append(desc)
            
            # Combine instruction and trials
            text_body = INSTRUCTION_TEXT + "\n".join(trials) + "\n"
            
            # Get optional metadata
            first_row = sub_df.iloc[0]
            
            result = {
                "participant_id": str(pid),
                "experiment": "devardaetal2024_rating",
                "text": text_body
            }
            
            # Add optional demographic metadata if available
            if pd.notna(first_row.get("age")):
                try:
                    result["age"] = int(first_row["age"])
                except (ValueError, TypeError):
                    pass
            if first_row.get("gender") and str(first_row["gender"]).strip():
                result["gender"] = str(first_row["gender"])
            if first_row.get("nationality") and str(first_row["nationality"]).strip():
                result["nationality"] = str(first_row["nationality"])
            if first_row.get("first_language") and str(first_row["first_language"]).strip():
                result["first_language"] = str(first_row["first_language"])
            if first_row.get("education") and str(first_row["education"]).strip():
                result["education"] = str(first_row["education"])
            
            fo.write(json.dumps(result, ensure_ascii=False) + "\n")
            records_written += 1
    
    print(f"Successfully wrote {records_written} participant records to: {OUTPATH}")
    
    # Create zip file
    with zipfile.ZipFile(ZIP_OUTPATH, 'w', zipfile.ZIP_DEFLATED) as zf:
        zf.write(OUTPATH, "prompts.jsonl")
    
    print(f"Created zip archive: {ZIP_OUTPATH}")
    
    # Remove unzipped file
    OUTPATH.unlink()
    print(f"Removed unzipped file: {OUTPATH}")
    
    # Verify token count for a sample participant
    sample_participant = participants[0]
    sample_text = df[df["participant_id"] == sample_participant]
    approx_tokens = len(INSTRUCTION_TEXT.split()) + sum(len(str(r).split()) for r in sample_text["stimulus"]) * 2
    print(f"\nApproximate tokens per participant: ~{approx_tokens} (well under 32K limit)")
    
    # Print example prompts
    print_example_prompts(df)


if __name__ == "__main__":
    main()
